### PR TITLE
Migrate: Postgres

### LINKED ISSUE #8: Port over to PostgreSQL

### PR DESCRIPTION
**Pull request by melon095**
This pull request aims to fully migrate the bot over to postgres.

**How to automatically migrate tables**
Use [PGLoader](https://github.com/dimitri/pgloader)
Will call the table `twitch`, rename to `bot`
Migration table -- update version to 2

Closes #8

### CODE REVIEW

**Comment by melon095** [comment]
All database calls should fit the new syntax of the `postgres` library.
Migrations deemed to have some issues as [Template String Functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates) moves every ${} to a seperate parameter, this means `postgres` will think this could be a *SELECT ?*, Where ? is the ${}.
The issue this created was that i can't call the query function which is *postgres\`FOO\`, because if i were to do postgres\`${migrationData}\`.

Luckely there is a file method, which will internally read a file.

File: package.json [hunk 1]
```diff
@@ -62,6 +62,7 @@
 		"got": "^12.0.2",
 		"humanize-duration": "^3.27.1",
 		"mysql2": "^2.3.3",
```
Comment (melon095) [inline package.json]: Remove this

File: src/controller/Banphrase/index.ts [hunk 2]
```diff
@@ -43,27 +43,21 @@ export class Banphrase {
...
+		});
 	}
 
 	async Update() {
-		Bot.SQL.promisifyQuery<PhraseType>(
-			'SELECT Phrase FROM `banphrases` WHERE `channel` = ?',
-			[this.Name],
-		)
-			.then((_p) => _p.ArrayOrNull())
+		Bot.SQL.Query<PhraseType[]>`SELECT Phrase FROM banphrases WHERE channel = ${this.Name}`
```
Comment (melon095) [inline src/controller/Banphrase/index.ts]: phrase

File: src/controller/Banphrase/index.ts [hunk 3]
```diff
@@ -43,27 +43,21 @@ export class Banphrase {
...
-		)
-			.then((_p) => _p.SingleOrNull())
-			.then((phrases) => {
-				if (phrases === null) return;
-				for (const phrase of phrases.Phrase) {
-					this.Bans.push(phrase);
-				}
-			});
+		Bot.SQL.Query<Database.banphrases[]>`
+            SELECT Phrase
```
Comment (melon095) [inline src/controller/Banphrase/index.ts]: phrase

File: src/controller/Commands/Handler.ts [hunk 4]
```diff
@@ -98,12 +88,7 @@ export class CommandsHandler {
 				for (const command of dir) {
 					if (command.name.split('.')[2] !== 'map') {
 						const name =
-							'file://' +
-							resolve(
-								process.cwd(),
-								'build/commands',
-								command.name,
-							);
+							'file://' + resolve(process.cwd(), 'build/commands', command.name);
```
Comment (melon095) [inline src/controller/Commands/Handler.ts]: tools.Import

File: src/controller/Commands/Handler.ts [hunk 5]
```diff
@@ -41,35 +41,25 @@ export class CommandsHandler {
...
-					command.Aliases?.includes(identifier),
+				(command) => command.Name === identifier || command.Aliases?.includes(identifier),
 			);
 
 			if (command === undefined) return Resolve(undefined);
 
-			const name =
-				'file://' +
-				resolve(process.cwd(), 'build/commands', command.Name + '.js');
+			const name = 'file://' + resolve(process.cwd(), 'build/commands', command.Name + '.js');
```
Comment (melon095) [inline src/controller/Commands/Handler.ts]: tools.Import

File: src/controller/DB/index.ts [hunk 6]
```diff
@@ -1,223 +1,73 @@
...
-				return this.Data[0] as unknown as T;
-			default:
-				// TODO: Should not do this?
-				throw new Error(
-					`SQL Data length is more than 1... Found length of: ${this.Data.length}`,
-				);
+        `.then(async ([row]) => {
+		if (row === undefined) {
+			await createDefaultMigrationTable(db);
+			db.Query`INSERT INTO migration (version) VALUES (0)`;
```
Comment (melon095) [inline src/controller/DB/index.ts]: await

File: src/controller/DB/index.ts [hunk 7]
```diff
@@ -226,39 +76,36 @@ export class SQLController {
...
-				'Migrations',
-				`${version}_${name}`,
-			);
-
-			const data = fs
-				.readFileSync(migrationPath, 'utf8')
-				.split(';')
-				.map((query) => query.trim())
-				.filter(Boolean);
+			// const migration = resolve(process.cwd(), 'Migrations', `${version}_${name}`);
```
Comment (melon095) [inline src/controller/DB/index.ts]: Remove comments

File: src/loops/loops.ts [hunk 8]
```diff
@@ -101,9 +88,10 @@ export function __loops() {
 
 						// Convert to string and convert back to array to join all array elements to one.
 						// const viewers: string = JSON.stringify(viewersArr.join(',').split(',')).replace("[", "").replace("]", "");
-						Bot.SQL.query(
-							'UPDATE `channels` SET `viewers` = JSON_ARRAY(?) WHERE `name` = ?',
-							[viewers.split(','), name],
+						Bot.SQL.Query(
```
Comment (melon095) [inline src/loops/loops.ts]: Parenthesis

**Review by melon095 — COMMENTED** [review]
Ok

**Review by melon095 — COMMENTED** [review]
Ok

File: config.example.json [hunk 9]
```diff
@@ -5,9 +5,7 @@
 		"ClientSecret": "" // https://dev.twitch.tv/
 	},
 	"SQL": {
-		"Host": "",
-		"User": "",
-		"Password": ""
+		"Address": "postgres://user:password@localhost:5432/melonbot"
```
Comment (hotbear1110) [inline config.example.json:8]: Hmm doesn't this depend on how the user has setup postgres?

Comment (melon095) [reply config.example.json:8]: Nah incase theres no password setup, im fairly sure you just omit it. So: `postgres://user@localhost:5432/foo`

Comment (hotbear1110) [reply config.example.json:8]: eShrug ok

**Review by hotbear1110 — APPROVED** [review]
Seems good I guess. Clueless not like I can actually test anything

**Comment by melon095** [comment]
image
Testing on Prod

**Comment by melon095** [comment]
image
Merging

**Pull request merged**